lib/index.js

```diff
@@ -72,388 +35,341 @@ const defaultRules = [
-      ? pkgFromPath(entry)
-      : null
-    const rootNM = isProject && entry === 'node_modules'
-    const rootPJ = isProject && entry === 'package.json'
-
-    return (
-      // if we're in a bundled package, check with the parent.
-      /^node_modules($|\/)/i.test(p) && !this.isProject ? this.parent.filterEntry(
-        this.basename + '/' + entry, partial)
-
-      // if package is bundled, all files included
-      // also include @scope dirs for bundled scoped deps
-      // they'll be ignored if no files end up in them.
-      // However, this only matters if we're in the root.
-      // node_modules folders elsewhere, like lib/node_modules,
-      // should be included normally unless ignored.
-      : pkg ? this.bundled.indexOf(pkg) !== -1 ||
-        this.bundledScopes.indexOf(pkg) !== -1
-
-      // only walk top node_modules if we want to bundle something
-      : rootNM ? !!this.bundled.length
-
-      // always include package.json at the root.
-      : rootPJ ? true
-
-      // always include readmes etc in any included dir
-      : packageMustHavesRE.test(entry) ? true
-
-      // npm-shrinkwrap and package.json always included in the root pkg
-      : isProject && (entry === 'npm-shrinkwrap.json' || entry === 'package.json')
-        ? true
-
-      // package-lock never included
-        : isProject && entry === 'package-lock.json' ? false
-
-        // otherwise, follow ignore-walk's logic
-        : super.filterEntry(entry, partial)
-    )
-  }
-
-  filterEntries () {
-    if (this.ignoreRules['.npmignore']) {
-      this.ignoreRules['.gitignore'] = null
+    // browser is required
+    if (browser) {
+      strict.push(`!/${browser}`)
     }
-    this.filterEntries = super.filterEntries
-    super.filterEntries()
-  }
 
-  addIgnoreFile (file, then) {
-    const ig = path.resolve(this.path, file)
-    if (file === 'package.json' && !this.isProject) {
-      then()
-    } else if (this.packageJsonCache.has(ig)) {
-      this.onPackageJson(ig, this.packageJsonCache.get(ig), then)
-    } else {
-      super.addIgnoreFile(file, then)
+    // main is required
+    if (main) {
+      strict.push(`!/${main}`)
     }
-  }
 
-  onPackageJson (ig, pkg, then) {
-    this.packageJsonCache.set(ig, pkg)
-
-    if (Array.isArray(pkg.files)) {
-      // in this case we already included all the must-haves
-      super.onReadIgnoreFile('package.json', pkg.files.map(
-        f => '!' + f
-      ).join('\n') + '\n', then)
-    } else {
-      // if there's a bin, browser or main, make sure we don't ignore it
-      // also, don't ignore the package.json itself, or any files that
-      // must be included in the package.
-      const rules = this.mustHaveFilesFromPackage(pkg).map(f => `!${f}`)
-      const data = rules.join('\n') + '\n'
-      super.onReadIgnoreFile(packageNecessaryRules, data, then)
+    // each bin is required
+    if (bin) {
+      for (const key in bin) {
+        strict.push(`!/${bin[key]}`)
+      }
     }
```

Comment:
You could stat these, but it probably doesn't hurt to add them even if they don't exist?

Reply:
yeah my logic was if you mention a file in a field that implies it's required we just blindly require it. a future pass could log warnings any time a file that looks to be required is missing, though